### PR TITLE
[DEV-5731] lazy prefixer proxy

### PR DESCRIPTION
**Pull request by gaetano-guerriero**
To be used as:
```
async def controller():
   prefixer_proxy  =LazyPrefixerProxy()
   awaith gather(a(prefixer_proxy), b(prefixer_proxy))

async def a(prefixer_proxy):
   if  cond_a:
      prefixer = prefixer_proxy()

async def b(prefixer_proxy):
   if  cond_b:
      prefixer = prefixer_proxy()
```

This change is [Reviewable](https://reviewable.io/reviews/athenianco/athenian-api/3252)

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@gaetano-guerriero We had exactly this code in the past: to lazily load the Prefixer asynchronously. As I remember, even under the same name! :smile: After several months of working with that proxy, I concluded that it was a bad idea and removed it. The reasons were:
- Higher complexity of the source code.
- More difficult to trace the performance. It could take a slot of more critical stuff, sometimes.
- All the places in the code needed the materialized Prefixer anyway.

So I suggest not repeating my mistake here. We will always have an async setup in each controller to load all the stuff: release settings, logical settings, etc. and the prefixer fits there very well.

File: server/athenian/api/controllers/dashboard_controller.py [hunk 1]
```diff
@@ -66,7 +65,8 @@ async def create_dashboard_chart(
     dashboard = await _get_request_dashboard(request, team_id, dashboard_id)
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/dashboard_controller.py:65]: for example, we should load the prefixer at this line, in the very first async call

File: server/athenian/api/controllers/goal_controller.py [hunk 2]
```diff
@@ -137,7 +137,7 @@ async def create_goal_template(request: AthenianWebRequest, body: dict) -> web.R
     await get_user_account_status_from_request(request, create_request.account)
 
     repositories = await parse_request_repositories(
-        create_request.repositories, request, create_request.account,
+        create_request.repositories, LazyPrefixerProxy(request, create_request.account),
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/goal_controller.py:140]: Also, if we remove the proxy, we will follow the same conventions that exist in the rest of the codebase, so it should help future programmers

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: server/requirements.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 aiohttp[speedups]==3.8.3
+async-lru==1.0.3
```
Comment (vmarkovtsev) [inline server/requirements.txt:2]: We will also avoid using this external package

File: server/athenian/api/internal/prefixer.py [hunk 4]
```diff
@@ -298,3 +299,20 @@ def __repr__(self) -> str:
...
+
+class LazyPrefixerProxy:
+    """When called return an instance of Prefixer, building it only when needed."""
+
+    def __init__(self, request, account):
+        """Init the `LazyPrefixerProxy`."""
+
+        @alru_cache
+        async def get_prefixer():
+            return await Prefixer.from_request(request, account)
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/prefixer.py:312]: Finally, this is likely not working as you'd expect. When you return an async function, it doesn't actually start to execute! You must wrap it into `asyncio.Task` to let it execute along with the rest of the code.

Also, we already have the memcached. It takes very few milliseconds to load the cached prefixed. So we won't win much here.

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: server/athenian/api/controllers/goal_controller.py (hunk 2 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/goal_controller.py:140]: Looking at the example, without proxy you can:

1 - always build prefixer, you do useless work; imagine cond_a or cond_b is True 10% of the time:
```
async def controller():
   prefixer = Prefixer()
   awaith gather(a(prefixer), b(prefixer))
async def a(prefixer):
   if  cond_a:
        prefixer.do_something()
async def b(prefixer):
   if  cond_b:
          prefixer.do_something()
```

2 - check conditions also in main function, propagate an optional prefixer, code gets complex, 
you repeat condition check, controller() has to mess with logic it shouldn't know, a() and b() has to mess with 
the assumption that the resource they need is there when they need it

```
async def controller():
   if cond_a or cond_b:
        maybe_prefixer = Prefixer()
   else:
        maybe_prefixer = None
   awaith gather(a(maybe_prefixer), b(maybe_prefixer))
async def a(prefixer):
   if  cond_a:
         assert prefixer is not None
        prefixer.do_something()
async def b(prefixer):
   if  cond_b:
          assert prefixer is not None
          prefixer.do_something()
```

3 - build prefixer twice, this is still better than 1 -  if cond_a or cond_b is False 10% of the time
```
async def controller():
   awaith gather(a(), b())
async def a():
   if  cond_a:
        prefixer = Prefixer()
        prefixer.do_something()
async def b():
   if  cond_b:
          prefixer = Prefixer()
          prefixer.do_something()
```

File: server/athenian/api/internal/prefixer.py (hunk 4 above)
Comment (gaetano-guerriero) [reply server/athenian/api/internal/prefixer.py:312]: :thinking: yes I don't want to execute it, it's lazy after all

File: server/athenian/api/controllers/goal_controller.py (hunk 2 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/goal_controller.py:140]: So I understand you suggest going for 1 - and trust memcached

**Pull request closed without merging**